crates/net/network-types/src/fetch_client.rs

```diff
@@ -0,0 +1,180 @@
+//! A client implementation that can interact with the network and download data.
+
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+use derive_more::Constructor;
+use futures::{future, future::Either};
+use reth_network_p2p::{
+    bodies::client::{BodiesClient, BodiesFut},
+    download::DownloadClient,
+    error::{PeerRequestResult, RequestError},
+    headers::client::{HeadersClient, HeadersRequest},
+    priority::Priority,
+};
+use reth_network_peers::PeerId;
+use reth_primitives::{BlockBody, Header, B256};
+use tokio::sync::{mpsc::UnboundedSender, oneshot};
+
+use crate::{FlattenedResponse, PeersHandle, ReputationChangeKind};
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// Front-end API for fetching data from the network.
+///
+/// Following diagram illustrates how a request, See [`HeadersClient::get_headers`] and
+/// [`BodiesClient::get_block_bodies`] is handled internally.
+///
+/// include_mmd!("docs/mermaid/fetch-client.mmd")
+#[derive(Debug, Clone, Constructor)]
+pub struct FetchClient {
```

Comment:
this doesn't really belong here imo, this is only really useful as an API type.

I'd like to not have this here

Reply:
sounds like we should have a `reth-network-primitives` crate, so we can keep using `reth-network-types` to avoid cyclical deps and reserve `reth-network-api` for traits, like we have done for `reth-rpc-eth-api`